### DISCUSSION TITLE
pyinstaller removers sqlalchemy modules

### DISCUSSION THREAD

**Discussion by aretaon** [sqlalchemy/sqlalchemy#10372] [Usage Questions] [answered]
I am trying to package a Python script that uses sqlalchemy with [mariadbconnector](https://pypi.org/project/mariadb/) into an exe with [pyinstaller](https://pypi.org/project/pyinstaller/). The code runs well from a standalone Python venv. My spec file includes 

`hiddenimports=['openpyxl', 'pywintypes', 'mariadb', 'sqlalchemy'],`

to capture sqlalchemy and mariadbconnector yet the packaged exe file fails with 

```
Traceback (most recent call last):
  File "PyBackupTask.py", line 1052, in <module>
  File "PyBackupTask.py", line 153, in test_connection
  File "<string>", line 2, in create_engine
  File "sqlalchemy\util\deprecations.py", line 298, in warned
  File "sqlalchemy\engine\create.py", line 522, in create_engine
  File "sqlalchemy\engine\url.py", line 636, in _get_entrypoint
  File "sqlalchemy\util\langhelpers.py", line 343, in load
sqlalchemy.exc.NoSuchModuleError: Can't load plugin: sqlalchemy.dialects:mariadb.mariadbconnector
```

I found that mariadbconnector relies on compiled code that is likely not included by pyinstaller. However, I don't know where to find that code to include it. I would be grateful if somebody could point me into the right direction.
Thanks!

**Comment by CaselIT** [comment]
> I found that mariadbconnector relies on compiled code that is likely not included by pyinstaller. 

from the issue it's likely not an issue with mariadbconnector.

Can you check that the sqlalchemy version used by pyinstaller is the same you use in the venv? it almost seems like that sqlalchemy version is an old one that does not explicitly support mariadb

**Reply by aretaon** [reply]
Thanks for the suggestion. I included a print statement for the package versions and they are the same in the (successful) run from the python console and in the (unsuccessful) run from the packaged exe file. 

Python:

> 2023-09-20 09:47:43,943 [MainThread  ] [INFO ]  Using sqlalchemy version 2.0.21.
> 2023-09-20 09:47:43,943 [MainThread  ] [INFO ]  Using mariadb version 1.1.7.

exe:

> 2023-09-20 09:49:17,943 [MainThread  ] [INFO ]  Using sqlalchemy version 2.0.21.
> 2023-09-20 09:49:17,943 [MainThread  ] [INFO ]  Using mariadb version 1.1.7.

**Reply by CaselIT** [reply]
that's strange. does pyinstaller has any sort of tree shaking feature? because from the error it seems that a module of sqlalchemy.dialects.mysql is not included in the exe for some reason, or is not found by python

**Reply by aretaon** [reply, marked as answer]
I solved the issue even though I am still not sure what the actual reason is (this is probably better addressed by the pyinstaller people):

The easiest fix is to include the import statement from `sqlalchemy.dialects.mysql.mariadb import *` directly in the Python script to package.
For convenience, I created a dedicated hook for that named `hook-sqlalchemy.dialects.mysql.mariadb.py` which contains nothing but this import statement. It is then referenced in the spec file by `runtime_hooks=['hook-sqlalchemy.dialects.mysql.mariadb.py']`.

Thanks @CaselIT for pointing me in direction of the mysql modules.

**Reply by CaselIT** [reply]
So it seems that pyinstall has tree shaking functionality and that was caused the issue.
It may make sense to make it import all the dialects you intend to use then
